kratos/containers/constraint_equation.h

```diff
@@ -0,0 +1,198 @@
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Aditya Ghantasala
+//
+//
+
+#if !defined(CONSTRAINT_EQUATION_H)
+#define CONSTRAINT_EQUATION_H
+// System includes
+#include <vector>
+#include <unordered_set>
+#include <iostream>
+#include <assert.h>
+
+// project includes
+#include "includes/define.h"
+#include "includes/dof.h"
+#include "includes/node.h"
+#include "includes/process_info.h"
+
+namespace Kratos
+{
+
+
+/**
+ * @class MasterSlaveRelation
+ * @ingroup KratosCore
+ * @brief This class stores the information regarding the MasterSlaveRelation equation. Naming convenction is defined like this. (each object of this class will store one equation in the given form
+ * @details *   SlaveEquationId = w_1*MasterEquationId_1 + w_2*MasterEquationId_2 + ..... + w_n*MasterEquationId_n
+ *
+ *   each slaveDOF and each of its masterDOF have the following attributes
+ *   a. dof ID
+ *   b. dof KEY
+ *   c. equation ID
+ *
+ *   one should be able to access the constraint equation both with (dofID , dofKEY) pair and/or equationID of the slave.
+ *   This equation is imposed on the linear system of equations either element wise or on the global system.
+ * @author Aditya Ghantasala
+ */
+class MasterSlaveRelation : public IndexedObject
+{
+  public:
+    typedef std::size_t IndexType;
+    typedef Matrix MatrixType;
+    typedef Vector VectorType;
+    typedef std::vector<IndexType> EquationIdVectorType;
+    KRATOS_CLASS_POINTER_DEFINITION(MasterSlaveRelation);
+
+    // empty constructor and methods to add master and slave independently.
+    MasterSlaveRelation() : IndexedObject(0), mSlaveEquationId(0)
+    {
+        SetConstant(0.0);
+        SetConstantUpdate(0.0);
+    }
+
+
+    MasterSlaveRelation(IndexType const &rSlaveEquationId) : IndexedObject(rSlaveEquationId), mSlaveEquationId(rSlaveEquationId)
+    {
+        SetConstant(0.0);
+        SetConstantUpdate(0.0);
+    }
+
+    void SetConstant(double Constant) { mConstant = Constant; }
+    void SetConstantUpdate(double ConstantUpdate) { mConstantUpdate = ConstantUpdate; }
+    double Constant() const { return mConstant; }
+    double ConstantUpdate() const { return mConstantUpdate; }
+    IndexType SlaveEquationId() const { return mSlaveEquationId; }
+
+
+
+    // Get number of masters for this slave
+    std::size_t GetNumberOfMasters() const
+    {
+        return mMasterDataSet.size();
+    }
+
+
+    /**
+     * this determines the master equation IDs connected to this constraint
+     * @param rResult the elemental equation ID vector
+     * @param rCurrentProcessInfo the current process info instance
+     */
+    virtual void EquationIdVector(EquationIdVectorType& rSlaveEquationIds,
+                                  EquationIdVectorType& rMasterEquationIds,
+                                  ProcessInfo& rCurrentProcessInfo)
+    {
+        if (rSlaveEquationIds.size() != 0)
+            rSlaveEquationIds.resize(1);
```

Comment:
this doesn't need to be 1...

Reply:
i mean you can connect M slaves to N masters by the corresponding matrix & vector